### ISSUE TITLE
Lua script - Enable HDR in Windows 10

### ISSUE THREAD

**Issue by Obegg** [mpv-player/mpv#10625]
Is it possible to create a .lua script that will enable HDR in Windows 10 if mpv detects `BT.2020` colorspace?

![image](https://user-images.githubusercontent.com/13978119/189238677-f83732ad-fd92-43a8-ad0b-f9dcc149d7fe.png)

I am aware of the fact that if I use 
```
vo=gpu-next
target-colorspace-hint=yes
gpu-api=vulkan
```

it will auto-enable HDR, without the need for a .lua script.
But, I don't want to use `gpu-api=vulkan`, I compared it to `gpu-api=d3d11` and the black levels were different (but that's a different issue),
So I would like to keep using `gpu-api=d3d11`.

In short, I would love a .lua script that will use logic like that:
If `bt.2020` colorspace is detected
then enable HDR in Windows 10 (it can be using pre-recorded mouse clicks or scripts or opening external program like AutoHotKey, etc)

Is something similar to this is possible to create?

**Comment by thebombzen** [comment]
Win+Alt+B is the keyboard shortcut to toggle HDR so you'd have to simulate that somehow.

**Comment by Obegg** [comment]
> Win+Alt+B is the keyboard shortcut to toggle HDR so you'd have to simulate that somehow.

I didn't know that, good to know, now I can use keyboard to toggle HDR, great!
I can simulate that if I have .lua script that will open an external program, problem is - I have no idea how to do that.

**Comment by thebombzen** [comment]
There's a `run` command that you can call with `mp.commandv()`

**Comment by Obegg** [comment]
I'm sorry, I have no idea how to use scripts, will you be able to help me with creating this .lua script so if it detects HDR it will run a test.exe from the same path as `scripts`?

**Comment by thebombzen** [comment]
Use auto-profiles.

**Comment by Obegg** [comment]
> Use auto-profiles.

By "auto-profiles" you mean `mpv.conf`?
Is `mpv.conf` able to run external program such as .exe? If so - what do I need to type?

**Comment by thebombzen** [comment]
> > Use auto-profiles.
> 
> By "auto-profiles" you mean `mpv.conf`? Is `mpv.conf` able to run external program such as .exe? If so - what do I need to type?

type `man mpv`

**Comment by Obegg** [comment]
Do you mean something like this?
```
[test]
profile-cond=p["frame-drop-count"]>1
man mpv
```
Because I get errors

**Comment by mitzsch** [comment]
I have written a Lua script and a small program that triggers HDR on windows, well all it does is simulate Win+Alt+B.

Script:
`toogleHDR.lua`
```
require "mp.msg"
require "mp.options"

mp.observe_property("video-params/primaries", "string", function (_, primaries)
    if primaries == nil then
        return
    end
    	
	if primaries == "bt.2020" then
	os.execute("C:\\path\\to\\enableHDR.exe")
	end
		
    print(mp.get_property("width"), "x", mp.get_property("height"), primaries)
	
	function offHDR()
        if primaries == "bt.2020" then
	       os.execute("C:\\path\\to\\enableHDR.exe")
	    end
	end

    mp.register_event("end-file", offHDR)
end)
```

`enableHDR.exe`
[enableHDR.zip](https://github.com/mpv-player/mpv/files/9532879/enableHDR.zip)
enableHDR.c
```
//
// enableHDR.c - Pauses, then simulates the shortcut to
// enable HDR in Windows 10 and 11 - Win + Alt + B
// 
//
// Originally written by Ted Burke 
// last updated 17-4-2012 - keystroke.c
//
// To compile with MinGW:
//
//      gcc -o keystroke.exe keystroke.c
//
// To run the program:
//
//      keystroke.exe
//
 
// Because the SendInput function is only supported in
// Windows 2000 and later, WINVER needs to be set as
// follows so that SendInput gets defined when windows.h
// is included below.
// #define WINVER 0x0500 - orig for Windows 2000
// set to minimum WINVER to Windows 10
#define WINVER 0x0A00
#include <windows.h>
 
int main()
{
    // This structure will be used to create the keyboard
    // input event.
    INPUT ip;
 
    // Pause for 5 seconds.
    // Sleep(5000);
 
    // Set up a generic keyboard event.
    ip.type = INPUT_KEYBOARD;
    ip.ki.wScan = 0; // hardware scan code for key
    ip.ki.time = 0;
    ip.ki.dwExtraInfo = 0;
 
    // Press the "WIN" key
    ip.ki.wVk = 0x5B; // virtual-key code for the "WIN" key
    ip.ki.dwFlags = 0; // 0 for key press
    SendInput(1, &ip, sizeof(INPUT));
	
	// Press the "ALT" key
    ip.ki.wVk = 0x12; // virtual-key code for the "ALT" key
    ip.ki.dwFlags = 0; // 0 for key press
    SendInput(1, &ip, sizeof(INPUT));
	
	// Press the "B" key
    ip.ki.wVk = 0x42; // virtual-key code for the "B" key
    ip.ki.dwFlags = 0; // 0 for key press
    SendInput(1, &ip, sizeof(INPUT));
 
    // Release the "WIN" key
	ip.ki.wVk = 0x5B;
    ip.ki.dwFlags = KEYEVENTF_KEYUP; // KEYEVENTF_KEYUP for key release
    SendInput(1, &ip, sizeof(INPUT));
	
	// Release the "ALT" key
	ip.ki.wVk = 0x12;
    ip.ki.dwFlags = KEYEVENTF_KEYUP; // KEYEVENTF_KEYUP for key release
    SendInput(1, &ip, sizeof(INPUT));
	
	// Release the "B" key
	ip.ki.wVk = 0x42;
    ip.ki.dwFlags = KEYEVENTF_KEYUP; // KEYEVENTF_KEYUP for key release
    SendInput(1, &ip, sizeof(INPUT));
 
    // Exit normally
    return 0;
}
```

**Comment by Obegg** [comment]
This is quite helpful, thank you!
Now all I need is to change to directory, this line:
```
os.execute("C:\\path\\to\\enableHDR.exe")
```
I need it to point to `%appdata%\mpv`, how can I do that without manually typing the full path?

Uh, and I also need to add a line that pause playback for 3 seconds or so.

**Comment by mitzsch** [comment]
> I need it to point to %appdata%\mpv, how can I do that without manually typing the full path?

I guess the easiest way to do so is actually manually typing in the path or copying the `.exe` to a simple place. (this is how I have done it)

**Comment by Obegg** [comment]
Is mpv able to detect the `primaries` of the display rather then the video?
Because if it is - It will be possible to create something with this logic:

```
if display primaries is bt.1886 and video primaries is bt.2020
run togglehdr.exe
if display is bt.2020 and video primaries is bt.1886
run togglehdr.exe
```

**Comment by ZaquL** [comment]
> if display primaries is bt.1886

Those are not primaries, this is a transfer and no, there is no way to know it is indeed not 2.200 gamma (technically, there is gamma in EDID, but that is never updated if you swicth to 2.4 gamma mode and besides BT.1886 is not pure 2.4 gamma).

**Comment by thebombzen** [comment]
It appears this has been solved.
